### DISCUSSION TITLE
N runner running multiple runnable at the same time

### DISCUSSION THREAD

**Discussion by marioaag** [avocado-framework/avocado#4498] [Q&A] [answered]
Could anybody explain me or point me to an example of how to run multiple runnable at the same time?

I can see that if the switch --test-runner='nrunner' is used avocado execute all the test in parallel but in someway the executions get isolated and looks like there is no way to specify the runnable values like the kind or the args.

In the other hand if I directly do it with the Python API like this:
`runnable = nrunner.Runnable('python-unittest', 'test')
        runner_klass = runnable.pick_runner_class()
        runner = runner_klass(runnable)
        results = [status for status in runner.run()]`

What is needed to do to get multiple runnable running at the same time? is it possible to pass multiple runnable to the same runner?

**Comment by richtja** [comment]
Hi @marioaag the runnables for each test are determinate by resolvers and avocado runs different kinds of tests based on the information from resolvers. Avocado will decide which runnable will be used for each test. So you can run multiple types of tests just by running them like this:
`avocado run  --test-runner=nrunner /bin/true examples/tests/passtest.py`

You will find more information in https://avocado-framework.readthedocs.io/en/latest/guides/contributor/chapters/runners.html#finding-tests

For having bigger power over test execution you can use Avocado API, but with API you have to define everything by yourself. You can find usage in our selftest https://github.com/avocado-framework/avocado/blob/master/selftests/check.py

Small example of setting runnable by Avocado API:

```
#!/usr/bin/env python3

import sys

from avocado.core.job import Job
from avocado.core.suite import TestSuite
from avocado.core.nrunner import Runnable

config = {
    'run.test_runner': 'nrunner',
    'core.show': ['app']
}

suite = TestSuite(name="sleep", config=config, tests=[Runnable('exec-test', '/bin/sleep', "3")])

with Job(config, [suite]) as j:
    sys.exit(j.run())
```

**Reply by marioaag** [reply]
Thanks this is a really good example!! it shows exactly what I want to do...

Now I know that there are couple of runnable kinds (noop, exec, exec-test, python-unittest) but not sure about the purpose of each one of them and if anyone of them could be used to run avocado tests written with Python created for the actual runner?

**Reply by clebergnu** [reply]
Hi @marioaag,

You should find some information in their docstrings, for instance:

https://github.com/avocado-framework/avocado/blob/9ea1274f84ecc068b097ff44d51707b5743baa31/avocado/core/nrunner.py#L355

https://github.com/avocado-framework/avocado/blob/9ea1274f84ecc068b097ff44d51707b5743baa31/avocado/core/nrunner.py#L374

https://github.com/avocado-framework/avocado/blob/9ea1274f84ecc068b097ff44d51707b5743baa31/avocado/core/nrunner.py#L433

https://github.com/avocado-framework/avocado/blob/9ea1274f84ecc068b097ff44d51707b5743baa31/avocado/core/nrunner.py#L465

Now, when you say "tests written with Python created for the actual runner", do you mean [Avocado's own instrumented tests](https://avocado-framework.readthedocs.io/en/86.0/guides/writer/chapters/writing.html#writing-avocado-tests-with-python)? If so, the matching kind is `avocado-instrumented`:

https://github.com/avocado-framework/avocado/blob/9ea1274f84ecc068b097ff44d51707b5743baa31/avocado/core/runners/avocado_instrumented.py#L12

Let me know if that helps.

**Reply by marioaag** [reply]
Thanks! that helps

Now the question is, that is for the lates version GitHub right? if I'm running in Avocado 86.0 that runner is not listed as part of the capabilities 

marioalvarado@MacBook-Pro avocado % avocado-runner capabilities
{"runnables": ["noop", "exec", "exec-test", "python-unittest"], "commands": ["capabilities", "runnable-run", "runnable-run-recipe", "task-run", "task-run-recipe"]}

is this a known issue?

Because looking a the source code I found that it is included but in a different place:

marioalvarado@MacBook-Pro avocado % ll core/nrunner_avocado_instrumented.py 
-rw-r--r--  1 marioalvarado  staff  4514 Mar 25 13:32 core/nrunner_avocado_instrumented.py

And reading the code of core/nrunner.py looks like it will find it if I define it as part of the runnable kind

**Reply by clebergnu** [reply, marked as answer]
That's because one runner script may implement support for more than one runnable kind, but it's not required that all runners be built into one script (such as `avocado-runner`).

For instance, on my machine I have a number of runner-scripts such as:

```
 ls -1 ~/.local/bin/avocado-runner*
/home/cleber/.local/bin/avocado-runner
/home/cleber/.local/bin/avocado-runner-avocado-instrumented
/home/cleber/.local/bin/avocado-runner-avocado-vt
/home/cleber/.local/bin/avocado-runner-exec
/home/cleber/.local/bin/avocado-runner-exec-test
/home/cleber/.local/bin/avocado-runner-golang
/home/cleber/.local/bin/avocado-runner-magic
/home/cleber/.local/bin/avocado-runner-noop
/home/cleber/.local/bin/avocado-runner-python-unittest
/home/cleber/.local/bin/avocado-runner-robot
/home/cleber/.local/bin/avocado-runner-tap
```

You may have fewer, but the general point here is that the nrunner architecture can cope with standalone runners for specific kinds.  They do not even need to be written in Python.

For an Avocado job to use one of those, it can either exist in the PATH or be registered as an entrypoint, such as here:

https://github.com/avocado-framework/avocado/blob/9ea1274f84ecc068b097ff44d51707b5743baa31/setup.py#L179
